### PR TITLE
Add support for uploading multiple files at once in file browser

### LINKED ISSUE #15198: Allow simultaneous multiple files upload
### Details

When creating a question, allow creator to simultaneously upload multiple images or files to clientFilesQuestion directory.

### PR DESCRIPTION
**Pull request by jonatanschroeder**
# Description

Resolves #15198. As the use of the Web UI becomes more prevalent, allowing instructors to upload multiple files at once becomes more useful. This PR adds support for uploading multiple files in the "Add files" and equivalent buttons in the file browser page. Uploading individual files (in the file viewer or in the Upload button for existing files) remains single-file.

- [x] I have disclosed the level of AI assistance used: Copilot used to help with instructions to migrate multer to multiple files. Codex 5.3 wrote the new unit test.

# Testing

Tested manually the related functionality, in multiple file browser contexts (both in root of individual context and in sub-directories), as well as with special directories. Also updated CI tests to include relevant tests.

Best reviewed hiding whitespaces.

### CODE REVIEW

File: apps/prairielearn/src/lib/config.ts [hunk 1]
```diff
@@ -176,6 +176,7 @@ export const ConfigSchema = z.object({
   sslKeyFile: z.string().default('/etc/pki/tls/private/localhost.key'),
   sslCAFile: z.string().default('/etc/pki/tls/certs/server-chain.crt'),
   fileUploadMaxBytes: z.number().default(1e7),
+  fileUploadMaxFiles: z.number().default(20),
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/config.ts:179]: I'm happy to brainstorm a different default limit here.

File: apps/prairielearn/src/lib/editors.ts [hunk 2]
```diff
@@ -2295,62 +2296,66 @@ export class FileUploadEditor extends Editor {
...
+      debug('oldHash: ' + oldHash);
+      debug('newHash: ' + newHash);
+      if (oldHash === newHash) {
+        debug('new contents are the same as old contents, so abort upload');
+        return false;
+      } else {
+        debug('new contents are different from old contents, so continue with upload');
+        return true;
+      }
+    });
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/editors.ts:2316]: To simplify reviewing, I'm moving this to a follow-up PR.

File: apps/prairielearn/src/lib/config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/config.ts:179]: Configs are managed by sysadmins, so strict validation is not necessary.

File: apps/prairielearn/src/lib/editors.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/editors.ts:2316]: This is drafted in #15216, testing pending.

File: apps/prairielearn/src/pages/instructorFileBrowser/instructorFileBrowser.ts [hunk 3]
```diff
@@ -162,30 +162,35 @@ router.post(
...
+              } catch {
+                throw new Error(
+                  `Invalid file path: ${req.body.working_path} / ${file.originalname}`,
+                );
+              }
+            }
+            return [filePath, file.buffer];
+          }),
+        );
+        const editor = new FileUploadEditor({ locals: res.locals, container, files });
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorFileBrowser/instructorFileBrowser.ts:203]: Done in 9b2068f.

**Comment by reteps** [comment]
Can I request that we merge #15004 first?

**Comment by jonatanschroeder** [comment]
> Can I request that we merge #15004 first?

Assuming you're concerned about merge conflicts, not a problem with me.

File: apps/prairielearn/src/tests/fileEditor.test.ts
Comment (EduardoMVAz) [inline apps/prairielearn/src/tests/fileEditor.test.ts:1]: Can you add a test for duplicate-name files?

File: apps/prairielearn/src/server.ts [hunk 4]
```diff
@@ -213,6 +213,7 @@ export async function initExpress(): Promise<Express> {
   const upload = multer({
     storage: multer.memoryStorage(),
     limits: {
+      files: config.fileUploadMaxFiles,
```
Comment (EduardoMVAz) [inline apps/prairielearn/src/server.ts:216]: For friendlier UX, when the user surpasses the file limit here, can we have an error that mirrors the "at most N files" message at the form?

File: apps/prairielearn/src/components/FileBrowser.tsx
Comment (EduardoMVAz) [inline apps/prairielearn/src/components/FileBrowser.tsx]: "Upload files"

Comment (EduardoMVAz) [inline apps/prairielearn/src/components/FileBrowser.tsx]: "Upload files"

**Review by EduardoMVAz — COMMENTED** [review]
Minor nits, but overall looks good. I would like to address these before merge, though.

**Comment by jonatanschroeder** [comment]
> > Can I request that we merge #15004 first?
> 
> Assuming you're concerned about merge conflicts, not a problem with me.

@EduardoMVAz do you still want to wait for this?

File: apps/prairielearn/src/server.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/server.ts:216]: Done in 9565b816a363a40cf8989bcdead447e382e46c6d. Also added validation for file size and duplicate files.

File: apps/prairielearn/src/components/FileBrowser.tsx
Comment (jonatanschroeder) [reply apps/prairielearn/src/components/FileBrowser.tsx]: 6e8b77cd2399a1d5e242ea6bbb67d96d8245e782

Comment (jonatanschroeder) [reply apps/prairielearn/src/components/FileBrowser.tsx]: 6e8b77cd2399a1d5e242ea6bbb67d96d8245e782

File: apps/prairielearn/src/tests/fileEditor.test.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/tests/fileEditor.test.ts:1]: 6b2e84395ba7306bdbd1455dd3302cf32031646b

**Comment by EduardoMVAz** [comment]
That will take longer than expected, given that Peter is no longer working with us. I think we can easily resolve the merge issues that come out of this.

**Review by EduardoMVAz — APPROVED** [review]
Nicely addressed.
